components/script/dom/servoparser/async_html.rs

```diff
@@ -0,0 +1,545 @@
+                }
+            }
+            ParseOperation::MarkScriptAlreadyStarted(node) => {
+                match self.nodes.borrow().get(&node) {
+                    Some(node) => {
+                        let script = node.downcast::<HTMLScriptElement>();
+                        script.map(|script| script.set_already_started(true));
+                    }
+                    None => panic!("Node not found!")
+                }
+            }
+            ParseOperation::CompleteScript(node) => {
+            }
+            ParseOperation::ReparentChildren(parent, new_parent) => {
+                let nodes = self.nodes.borrow();
+                let parent = nodes.get(&parent).unwrap();
+                let new_parent = nodes.get(&new_parent).unwrap();
+                while let Some(ref child) = parent.GetFirstChild() {
+                    new_parent.AppendChild(&child).unwrap();
+                }
+            }
+            ParseOperation::AssociateWithForm(target, form) => {
+                let nodes = self.nodes.borrow();
+                let node = nodes.get(&target).unwrap();
+                let form = nodes.get(&form).unwrap();
+
+                let form = Root::downcast::<HTMLFormElement>(Root::from_ref(&**form))
+                    .expect("Owner must be a form element");
+
+                let elem = node.downcast::<Element>();
+                let control = elem.and_then(|e| e.as_maybe_form_control());
+
+                if let Some(control) = control {
+                    control.set_form_owner_from_parser(&form);
+                } else {
+                    // TODO remove this code when keygen is implemented.
+                    assert!(node.NodeName() == "KEYGEN", "Unknown form-associatable element");
+                }
+            }
+            ParseOperation::Pop(node) => {
+                let nodes = self.nodes.borrow();
+                let node = nodes.get(&node).unwrap();
+                let node = Root::from_ref(&**node);
+                vtable_for(&node).pop();
+            }
+            ParseOperation::CreatePI(node) => {
+                let parse_node_data = self.parse_node_data.borrow();
+                let data = parse_node_data.get(&node).unwrap();
+                let pi = ProcessingInstruction::new(
+                    DOMString::from(data.target.clone().unwrap()),
+                    DOMString::from(data.data.clone().unwrap()),
+                    document);
+                self.nodes.borrow_mut().insert(node, JS::from_ref(pi.upcast()));
+            }
+        }
+    }
+}
+
+#[allow(unrooted_must_root)]
+impl TreeSink for Sink {
+    type Output = Self;
+    fn finish(self) -> Self { self }
+
+    type Handle = ParseNode;
+
+    fn get_document(&mut self) -> Self::Handle {
+        ParseNode {
+            id: 0
+        }
+    }
+
+    fn get_template_contents(&mut self, target: &Self::Handle) -> Self::Handle {
+        let contents = self.new_parse_node();
```

Comment:
That means every time you get the template contents, a new parse node is created. That is wrong, the same one should be returned every time.